feat: create symlink operator

rinzool opened the pull request:
## Why
See: https://github.com/treeverse/airflow-provider-lakeFS/issues/40 

## How
* Update hook to add a `create_symlink_file` method
* Create a `CreateSymlinkOperator` which takes a repo and a branch and creates a symlink file 

## Tests
I validated it locally by doing:
* A Airflow connection to my lakefs endpoint
* A simple lakefs repository with one file
* One dag with one sing `LakeFSCreateSymlinkOperator` 
```python
from airflow.decorators import dag
from airflow.utils.dates import days_ago

from lakefs_provider.operators.create_symlink_opeartor import LakeFSCreateSymlinkOperator


@dag(
    dag_id="poc_lakefs",
    default_args={
        "retries": 0,
    },
    schedule_interval=None,
    start_date=days_ago(1),
)
def create_dag():
    LakeFSCreateSymlinkOperator(lakefs_conn_id="conn_lakefs", task_id="create_symlink", repo="datamarts", branch="main")

DAG = create_dag()
```
* Run the dag
* Validate that symlink was created and task in success

I also updated `example-dags/lakefs-dag.py` to add symlink file creation on `example-branch` branch. When running the DAG locally using astro, we can see that symlink file was created by running
```
$ docker-compose -f ops/docker-compose.yaml exec lakefs ls lakefs/data/block/data/symlinks/example-repo/example-branch/path/to/symlink.txt
```
And its content
```
$ docker-compose -f ops/docker-compose.yaml exec lakefs cat lakefs/data/block/data/symlinks/example-repo/example-branch/path/to/symlink.txt
local://data/data/gqg5v29nnvsc739npc10/cena8f9nnvsc739npc30
```

Jonathan-Rosenberg reviewed (changes requested):
Very cool, thank you for that @rinzool!

Would you mind adding the optional [`location` parameter](https://pydocs.lakefs.io/docs/MetadataApi.html#create_symlink_file)?
https://docs.lakefs.io/reference/api.html#/metadata/createSymlinkFile

rinzool commented:
> Very cool, thank you for that @rinzool!
> 
> Would you mind adding the optional [`location` parameter](https://pydocs.lakefs.io/docs/MetadataApi.html#create_symlink_file)? https://docs.lakefs.io/reference/api.html#/metadata/createSymlinkFile

Done! Thank you I didn't notice that `location` was an optional parameter of the route

Furthermore, I was considering defining a default value for `lakefs_conn_id` like `lakefs_conn_id = "conn_lakefs"` (like other providers such as databricks or airflow which define default value for conn id)
But it is not the case for other operators in this provider so I wasn't sure

Jonathan-Rosenberg commented on lakefs_provider/hooks/lakefs_hook.py:
```diff
@@ -124,3 +124,7 @@ def stat_object(self, repo: str, ref: str, path: str) -> ObjectStats:
         client = self.get_conn()
         return client.objects.get_object(repository=repo, ref=ref, path=path)
+
+    def create_symlink_file(self, repo: str, branch: str)  -> str:
```
Where do you address the optional `location` property?

Jonathan-Rosenberg commented on lakefs_provider/operators/create_symlink_operator.py:
```diff
@@ -0,0 +1,60 @@
+        lakefs_conn_id: str,
+        repo: str,
+        branch: str,
+        location: str = None,
```
You don't set it anywhere...

Jonathan-Rosenberg commented on lakefs_provider/operators/create_symlink_operator.py:
```diff
@@ -0,0 +1,60 @@
+            self.log.info("Create symlink in location '%s'", location)
+            api_kwargs["location"] = location
+
+        location = hook.create_symlink_file(self.repo, self.branch, **api_kwargs)
```
Please use a different name for the location so it will reflect the storage URI location and not the optional param.

Jonathan-Rosenberg reviewed (changes requested):
Thanks!
Just a few comments in the files themselves...
Regarding `lakefs_conn_id`, sounds like a good idea yet not in the context of this PR. This will also imply updating the documentation.

rinzool commented:
Sorry @Jonathan-Rosenberg, now it should be all good!

lukland commented on lakefs_provider/operators/create_symlink_operator.py:
```diff
@@ -0,0 +1,56 @@
+
+class LakeFSCreateSymlinkOperator(BaseOperator):
+    """
+    Create a symlnik file
```
```suggestion
    Create a symlink file
```

lukland commented on lakefs_provider/operators/create_symlink_operator.py:
```diff
@@ -0,0 +1,56 @@
+
+from airflow.models import BaseOperator
+from airflow.utils.decorators import apply_defaults
+
```
```suggestion
```
In order to be [PEP8 compliant](https://peps.python.org/pep-0008/#imports)

lukland reviewed (commented):
Hello, I was passing by and saw a few things to fix.

arielshaqed commented:
Hi @rinzool ,

Thanks for the PR. Really excited to see serious usage increasing! 🥳 

One tricky thing: It should eventually have tests. :slightly_smiling_face: Otherwise things break.

The most important thing would be a description (in a comment on the PR) about how you checked this.

@itaiad200  did some stuff based on how Astronomer write tests, so [this bit](https://github.com/treeverse/airflow-provider-lakeFS/blob/0cd7151a800628a228eb9762bdaae607e687126a/.github/workflows/provider.yaml#L59-L65) runs them and [this is the entrypoint](https://github.com/treeverse/airflow-provider-lakeFS/blob/0cd7151a800628a228eb9762bdaae607e687126a/lakefs_provider/example_dags/lakefs-dag.py#L75) to the test.  If it's too difficult, let's pull and I can probably add a test myself. But I would need a description of the manual test you performed: setup, the DAG you ran, the result.  I would like the test to be able verify that we don't break your use case!

Thanks again!

rinzool commented:
Thanks @arielshaqed 
I updated `lakefs-dag.py` to add the operator on `example-branch` and I updated `Provider` workflow so it checks that symlink file exists and verifies that it has one line containing `local://`

I executed all workflow commands locally and it works fine (the DAG has the new operator, and I can see the symlink). Can you please approve the workflow please? :pray: Note that the [previous run](https://github.com/treeverse/airflow-provider-lakeFS/actions/runs/3799287216/jobs/6468120674) failed with a 403 when running command
```AWS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE AWS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY aws s3 cp --endpoint-url=http://s3.local.lakefs.io:8000/ s3://example-repo/main/path/to/_SUCCESS -```
I don't really understand why (it seems not related to my PR but I'm not sure) :thinking:

rinzool commented:
@arielshaqed And about testing, while developing this feature, I was thinking about adding some simple unit tests with pytest (or unittest) to validate that my provider calls the good hook method with appropriate arguments, etc. But I did not feel legitimate to add unit tests with a workflow on a repository that I do not own

But I just commited a _pytest_ workflow and a unit test files for my new operator. If necessary I will drop the commit :wink:

arielshaqed commented:
Hi @rinzool!

First of all, thanks so much for adding tests -- it makes it much easier to understand the intent of the code, and even more importantly it will allow us to ensure that your contribution continues to work with as lakeFS (and even Amazonian S3 deployments...) change!

I see we are running into problems getting the **unit test** to work.  These problems are not related to code quality, rather there is some issue (that I do not understand...) with Airflow and/or Python and/or GitHub actions and/or something else integration.  I do not want to block on getting the unit test to work: it tests that your code works in isolation, when in fact the bad failure modes I can imagine are in integration (incompatibility with the lakeFS API, bad composability in Airflow, ...).

I do not want to block this PR on the unit test - please feel very free to remove it entirely.

On the other hand, for precisely these reasons I want us to get the integration test to pass!  Unfortunately that means that I cannot approve until does.

I will work with you until it passes; this may end up with some small fixes of mine, so I'd be happy to know you won't mind me pushing here.

rinzool commented:
Thanks @arielshaqed for helping me, please feel free to push on this branch! I managed to fix the unit test workflow: there was an issue when installing requirements:
```
[launchpadlib 1.10.6 requires testresources which is not installed](https://gist.github.com/y56/0540d22a1db40dacc7fbbb93c866821e)``` (issue described [here](https://gist.github.com/y56/0540d22a1db40dacc7fbbb93c866821e))
I fixed it by using the latest ubuntu runner.
I hoped that it would fix the _Provider_ workflow but unfortunately not :confused:

arielshaqed commented:
I believe the integration ("provider") test will start working once #43 is done, and then we'll pull this.

arielshaqed commented:
@rinzool could you rebase on main (now with #43)?  I believe that will allow the integration test at least to _start_.

@Jonathan-Rosenberg did a much more in-depth review of the code and still has changes marked requested.

Jonathan-Rosenberg reviewed (approved):
LGTM
Thanks @rinzool !

arielshaqed commented:
Tests pass, neat!  Thanks for bearing with us for this, and see you on the lake!
